### PR TITLE
Fix JSON mapping linting against subset of builtin types

### LINKED ISSUE #11072: JSON field mapping validation
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.17

#### Summary

After #10946 when linting entities started getting:

> The field 'App\Entity\Entity#value' has the property type 'mixed' that differs from the metadata field type 'array' returned by the 'json' DBAL type.
> The field 'App\Entity\AnotherEntity#value' has the property type 'int' that differs from the metadata field type 'array' returned by the 'json' DBAL type.

JSON is not only array, but also can be a string or a number.

#### Current behavior

Wrong validation

#### How to reproduce

Declare json columns with other than array type.

#### Expected behavior

No errors

### PR DESCRIPTION
**Pull request by norkunas**
Closes #11072

### CODE REVIEW

**Review by greg0ire — DISMISSED** [review]

**Review by SenseException — DISMISSED** [review]

File: lib/Doctrine/ORM/Tools/SchemaValidator.php [hunk 1]
```diff
@@ -406,6 +406,13 @@ function (array $fieldMapping) use ($class): ?string {
                             );
                         }
 
+                        if (
+                            $fieldMapping['type'] === 'json'
+                            && in_array($propertyType, ['string', 'int', 'float', 'bool', 'true', 'false', 'null'], true)
```
Comment (janatjak) [inline lib/Doctrine/ORM/Tools/SchemaValidator.php:411]: Can you please add `mixed` ?

example
```php
    /** @Column(type="json") */
    public mixed $data = null;
```

Comment (janatjak) [reply lib/Doctrine/ORM/Tools/SchemaValidator.php:411]:
```suggestion
                            && in_array($propertyType, ['string', 'int', 'float', 'bool', 'true', 'false', 'mixed', 'null'], true)
```

Comment (norkunas) [reply lib/Doctrine/ORM/Tools/SchemaValidator.php:411]: > Can you please add `mixed` ?
> 
> example
> 
> ```
>     /** @Column(type="json") */
>     public mixed $data = null;
> ```

sure, added

**Comment by derrabus** [comment]
How does this fix play with union types? For example:

```php
#[Column(type: 'json')]
private array|bool|null $data;
```

**Comment by norkunas** [comment]
> How does this fix play with union types? For example:
> 
> ```
> #[Column(type: 'json')]
> private array|bool|null $data;
> ```

It is not linted, because it's already ignored here:
https://github.com/doctrine/orm/blob/93d205797bc0bba76861ae4f50566c3194d4e75e/lib/Doctrine/ORM/Tools/SchemaValidator.php#L373-L376

**Comment by derrabus** [comment]
Ah, perfect. 😎

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH11072/GH11072EntityBasic.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public float $score = 0.0;
+
+    /** @Column(type="json", nullable=true) */
+    public ?bool $trinary = null;
+
+    /** @Column(type="json") */
+    public array $metadata = [];
+
+    /** @Column(type="json") */
+    public mixed $anything;
```
Comment (derrabus) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH11072/GH11072EntityBasic.php]: This won't work on PHP 7. Let's move it to `GH11072EntityAdvanced`?

Comment (norkunas) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH11072/GH11072EntityBasic.php]: If it's ok for you to run it only on 8.2 then I could, but locally now added a intermediary for the `mixed` type which would run on >= 8.0

Comment (derrabus) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH11072/GH11072EntityBasic.php]: > If it's ok for you to run it only on 8.2

I think this is fine.

**Comment by derrabus** [comment]
> It is not linted, because it's already ignored here:
> 
> https://github.com/doctrine/orm/blob/93d205797bc0bba76861ae4f50566c3194d4e75e/lib/Doctrine/ORM/Tools/SchemaValidator.php#L373-L376

Actually, I think we should extend this condition to also include `mixed`. Performing this validation on `mixed` properties is pretty pointless, given that `mixed` is a union of every value type known to PHP. We're about to build a special case for JSON on `mixed`, but I think this applies to more or less any DBAL type, built-in or custom.

**Comment by norkunas** [comment]
Updated to always ignore mixed type

**Comment by norkunas** [comment]
Can we get this merged? :slightly_smiling_face:

**Review by greg0ire — APPROVED** [review]

**Review by derrabus — APPROVED** [review]

**Comment by norkunas** [comment]
Thank you

**Comment by gnutix** [comment]
@derrabus @greg0ire Would you mind releasing a 2.17.2 version with this fix ? :pray:

**Pull request merged**